projects/l2_pooling/noise_tolerance_l2_l4.py

```diff
@@ -0,0 +1,357 @@
+  """
+  n = int(len(pattern) * (1.0 - dropoutPercent))
+
+  return set(random.sample(pattern, n))
+
+
+def createRandomObjects(numObjects, locationsPerObject, featurePoolSize):
+  """
+  Generate random objects.
+
+  @param numObjects (int)
+  The number of objects to generate
+
+  @param locationsPerObject (int)
+  The number of points on each object
+
+  @param featurePoolSize (int)
+  The number of possible features
+
+  @return
+  For example, { 0: [0, 1, 2],
+                 1: [0, 2, 1],
+                 2: [2, 0, 1], }
+  is 3 objects. The first object has Feature 0 and Location 0, Feature 1 at
+  Location 1, Feature 2 at location 2, etc.
+  """
+
+  allFeatures = range(featurePoolSize)
+  allLocations = range(locationsPerObject)
+  objects = dict((name,
+                     [random.choice(allFeatures) for _ in xrange(locationsPerObject)])
+                    for name in xrange(numObjects))
+
+  return objects
+
+
+
+def doExperiment(numColumns, objects, l2Overrides, noiseLevels, numInitialTraversals,
+                 featureDropout, locationDropout):
+  """
+  Touch every point on an object 'numInitialTraversals' times, then evaluate
+  whether it has inferred the object by touching every point once more and
+  checking the number of correctly active and incorrectly active cells.
+
+  @param numColumns (int)
+  The number of sensors to use
+
+  @param l2Overrides (dict)
+  Parameters for the ColumnPooler
+
+  @param objects (dict)
+  A mapping of object names to their features.
+  See 'createRandomObjects'.
+
+  @param noiseLevels (list of floats)
+  The noise levels to experiment with. The experiment is run once per noise
+  level. Noise is applied at a constant rate to exactly one cortical column.
+  It's applied to the same cortical column every time, and this is the cortical
+  column that is measured.
+
+  @param featureDropout (bool)
+  Whether to use a noisy feature
+
+  @param locationDropout (bool)
+  Whether to use a noisy location
+  """
+
+  featureSDR = lambda : set(random.sample(xrange(2048), 40))
+  locationSDR = lambda : set(random.sample(xrange(1024), 40))
+
+  featureSDRsByColumn = [defaultdict(featureSDR) for _ in xrange(numColumns)]
+  locationSDRsByColumn = [defaultdict(locationSDR) for _ in xrange(numColumns)]
+
+  exp = L4L2Experiment(
+    "Experiment",
+    numCorticalColumns=numColumns,
+    externalInputSize=1024,
+    seed=random.randint(2048, 4096)
+  )
+
+  exp.learnObjects(
+    dict((objectName,
+          [dict((column,
+                 (locationSDRsByColumn[column][location],
+                  featureSDRsByColumn[column][features[location]]))
+                for column in xrange(numColumns))
+           for location in xrange(len(features))])
+         for objectName, features in objects.iteritems()))
+
+  if locationDropout:
+    # Now that the objects are learned, allow bursting L4 minicolumns to
+    # activate cells in L2.
+    exp.L4Columns[0].setParameter("defaultOutputType", 0, "active")
```

Comment:
This is a good point. At first I agreed with you, but I think we actually we want to isolate this behavior to the noise tolerance tests for now, and leave the other scripts unaffected.